notificationworkerlambda/src/main/scala/com/gu/notifications/worker/delivery/fcm/oktransport/OkGoogleHttpRequest.scala

```diff
@@ -0,0 +1,100 @@
+package com.gu.notifications.worker.delivery.fcm.oktransport
+
+import java.io.ByteArrayInputStream
+import java.util.concurrent.ConcurrentLinkedQueue
+
+import com.google.api.client.http.{LowLevelHttpRequest, LowLevelHttpResponse}
+import okhttp3.{Headers, MediaType, OkHttpClient, Request, RequestBody, Response}
+import okio.BufferedSink
+
+import scala.annotation.tailrec
+import scala.collection.mutable.ArrayBuffer
+
+
+class OkGoogleHttpRequest(okHttpClient: OkHttpClient, url: String, method: String) extends LowLevelHttpRequest {
+  //No idea whether FCM now or in the future will call addHeaders concurrently
+  private val headers = new ConcurrentLinkedQueue[(String, String)]()
+
+  override def addHeader(name: String, value: String): Unit = headers.offer((name, value))
+
+  override def execute(): LowLevelHttpResponse = {
+    val response: Response = executeRequest()
+    val (maybeContentType: Option[String], maybeContentLength: Option[Long], maybeContent: Option[ByteArrayInputStream]) = readBody(response)
+    val headerList: scala.List[(String, String)] = readHeaders(response)
+    val maybeContentEncoding = Option(response.header("content-encoding"))
+    val httpProtocolInUpperCase = response.protocol().toString.toUpperCase
+    val statusCode = response.code
+    val statusReason = Option(response.message()).map(_.trim).filterNot(_.isEmpty)
+    new OkGoogleHttpResponse(
+      maybeContent,
+      maybeContentEncoding,
+      maybeContentLength,
+      maybeContentType,
+      httpProtocolInUpperCase,
+      statusCode,
+      statusReason,
+      headerList
+    )
+  }
+
+  private def readHeaders(response: Response): List[(String, String)] = {
+    val headerTuples = ArrayBuffer[(String, String)]()
+    val okHeaders: Headers = response.headers()
+    val headerCount = okHeaders.size()
+    for (headerIndex <- 0 until headerCount) {
+      headerTuples += ((okHeaders.name(headerIndex), okHeaders.value(headerIndex)))
+    }
+    headerTuples.toList
+  }
+
+  private def readBody(response: Response): (Option[String], Option[Long], Option[ByteArrayInputStream]) = {
```

Comment:
It looks like the reason you return these three attribute together is because you want to close the body after.

What I suggest would be to have a function defined like so:

```scala
private def withResponseBody[A](response: response)(f: Body => A): A {
  val body = Option(response.body())
  val result = body.map(f)
  body.foreach(_.close())
  result
}
```

then you can pick whatever attributes you want from the body and that allows you to keep the truple definition local to the function that uses it

```scala
val (maybeContentType: Option[String], maybeContentLength: Option[Long], maybeContent: Option[ByteArrayInputStream]) = withResponseBody(response) { body =>
  (Option(body.getContentType), Option(...))
}

```

Reply:
Done